optuna/terminator/improvement/evaluator.py

```diff
@@ -64,53 +87,98 @@ class RegretBoundEvaluator(BaseImprovementEvaluator):
 
     def __init__(
         self,
-        gp: Optional[BaseGaussianProcess] = None,
         top_trials_ratio: float = DEFAULT_TOP_TRIALS_RATIO,
         min_n_trials: int = DEFAULT_MIN_N_TRIALS,
-        min_lcb_n_additional_samples: int = 2000,
+        seed: int | None = None,
     ) -> None:
-        self._gp = gp or _BoTorchGaussianProcess()
         self._top_trials_ratio = top_trials_ratio
         self._min_n_trials = min_n_trials
-        self._min_lcb_n_additional_samples = min_lcb_n_additional_samples
-
-    def get_preprocessing(self, add_random_inputs: bool = False) -> BasePreprocessing:
-        processes = [
-            SelectTopTrials(
-                top_trials_ratio=self._top_trials_ratio,
-                min_n_trials=self._min_n_trials,
-            ),
-            UnscaleLog(),
-            ToMinimize(),
-        ]
-
-        if add_random_inputs:
-            processes += [AddRandomInputs(self._min_lcb_n_additional_samples)]
-
-        processes += [OneToHot()]
-
-        return PreprocessingPipeline(processes)
+        self._log_prior = prior.default_log_prior
+        self._minimum_noise = prior.DEFAULT_MINIMUM_NOISE_VAR
+        self._optimize_n_samples = 2048
+        self._rng = LazyRandomState(seed)
 
     def evaluate(
```

Comment:
I would code like this to narrow down the variable scopes and enhance the readability.
IMO, the calculation of standardized regret bound should be separated from the main method.

```python
    def _compute_standardized_regret_bound(
        self,
        kernel_params: "gp.KernelParamsTensor",
        gp_search_space: "search_space.SearchSpace",
        normalized_params: np.ndarray,
        standardized_scores: np.ndarray,
    ) -> float:
        """
        In the original paper, f(x) was intended to be minimized, but here we would like to
        maximize f(x), i.e. score_vals. Hence, the following changes happen:
            1. min(ucb) over top trials becomes max(lcb) over top trials, and
            2. min(lcb) over the search space becomes max(ucb) over the search space, and
            3. Regret bound becomes max(ucb) over the search space minus max(lcb) over top trials.
        """
        n_trials, n_params = normalized_params.shape
        sqrt_beta = np.sqrt(_get_beta(n_params=n_params, n_trials=n_trials))
        ucb_acqf_params = acqf.create_acqf_params(
            acqf_type=acqf.AcquisitionFunctionType.UCB,
            kernel_params=kernel_params,
            search_space=gp_search_space,
            X=normalized_params,
            Y=standardized_scores,
            sqrt_beta=sqrt_beta,
        )
        seed = self._rng.rng.randint(np.iinfo(np.int32).max)
        # UCB over the search space. (Original: LCB over the search space. See Change 1 above.)
        max_ucb = np.max(
            acqf.eval_acqf_no_grad(ucb_acqf_params, normalized_params),
            initial=optim.optimize_acqf_sample(ucb_acqf_params, self._optimize_n_samples, seed)[1],
        )
        lcb_acqf_params = acqf.create_acqf_params(
            acqf_type=acqf.AcquisitionFunctionType.UCB,
            kernel_params=kernel_params,
            search_space=gp_search_space,
            X=normalized_params,
            Y=standardized_scores,
            sqrt_beta=-sqrt_beta,
        )
        # LCB over the top trials. (Original: UCB over the top trials. See Change 2 above.)
        max_lcb = np.max(acqf.eval_acqf_no_grad(lcb_acqf_params, normalized_params))
        # max(UCB) - max(LCB). (Original: min(UCB) - min(LCB). See Change 3 above.)
        return max_ucb - max_lcb

    def _get_top_n(
        self, params: np.ndarray, score_vals: np.ndarray
    ) -> tuple[np.ndarray, np.ndarray]:
        assert len(params) == len(score_vals)
        n_trials = len(params)
        top_n = np.clip(int(n_trials * self._top_trials_ratio), self._min_n_trials, n_trials)
        top_n_val = np.partition(score_vals, n_trials - top_n)[n_trials - top_n]
        top_mask = score_vals >= top_n_val
        return params[top_mask], score_vals[top_mask]

    def evaluate(self, trials: List[FrozenTrial], study_direction: StudyDirection) -> float:
        optuna_search_space = intersection_search_space(trials)
        self._validate_input(trials, optuna_search_space)
        complete_trials = [t for t in trials if t.state == TrialState.COMPLETE]

        # _gp module assumes that optimization direction is maximization
        sign = -1 if study_direction == StudyDirection.MINIMIZE else 1
        score_vals = np.array([t.value for t in complete_trials]) * sign
        gp_search_space, normalized_params = search_space.get_search_space_and_normalized_params(
            complete_trials, optuna_search_space
        )

        top_n_normalized_params, top_n_scores = self._get_top_n(normalized_params, score_vals)
        top_n_score_std = max(1e-10, top_n_scores.std())
        top_n_standardized_scores = (top_n_scores - np.mean(top_n_scores)) / top_n_score_std
        kernel_params = gp.fit_kernel_params(
            X=top_n_normalized_params,
            Y=top_n_standardized_scores,
            is_categorical=(gp_search_space.scale_types == search_space.ScaleType.CATEGORICAL),
            log_prior=self._log_prior,
            minimum_noise=self._minimum_noise,
            initial_kernel_params=None,
        )
        standardized_regret = self._compute_standardized_regret_bound(
            kernel_params=kernel_params,
            gp_search_space=gp_search_space,
            normalized_params=top_n_normalized_params,
            standardized_scores=top_n_standardized_scores,
        )
        return standardized_regret * top_n_score_std
```